lib/blockstore/badger/blockstore.go

```diff
@@ -0,0 +1,416 @@
+		return ErrBlockstoreClosed
+	}
+
+	batch := b.DB.NewWriteBatch()
+	defer batch.Cancel()
+
+	// toReturn tracks the byte slices to return to the pool, if we're using key
+	// prefixing. we can't return each slice to the pool after each Set, because
+	// badger holds on to the slice.
+	var toReturn [][]byte
+	if b.prefixing {
+		toReturn = make([][]byte, 0, len(blocks))
+		defer func() {
+			for _, b := range toReturn {
+				KeyPool.Put(b)
+			}
+		}()
+	}
+
+	for _, block := range blocks {
+		k, pooled := b.PooledStorageKey(block.Cid())
+		if pooled {
+			toReturn = append(toReturn, k)
+		}
+		if err := batch.Set(k, block.RawData()); err != nil {
+			return err
+		}
+	}
+
+	err := batch.Flush()
+	if err != nil {
+		err = fmt.Errorf("failed to put blocks in badger blockstore: %w", err)
+	}
+	return err
+}
+
+// DeleteBlock implements Blockstore.DeleteBlock.
+func (b *Blockstore) DeleteBlock(cid cid.Cid) error {
+	if atomic.LoadInt64(&b.state) != stateOpen {
+		return ErrBlockstoreClosed
+	}
+
+	k, pooled := b.PooledStorageKey(cid)
+	if pooled {
+		defer KeyPool.Put(k)
+	}
+
+	return b.DB.Update(func(txn *badger.Txn) error {
+		return txn.Delete(k)
+	})
+}
+
+// AllKeysChan implements Blockstore.AllKeysChan.
+func (b *Blockstore) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
+	if atomic.LoadInt64(&b.state) != stateOpen {
+		return nil, ErrBlockstoreClosed
+	}
+
+	txn := b.DB.NewTransaction(false)
+	opts := badger.IteratorOptions{PrefetchSize: 100}
+	if b.prefixing {
+		opts.Prefix = b.prefix
+	}
+	iter := txn.NewIterator(opts)
+
+	ch := make(chan cid.Cid)
+	go func() {
+		defer close(ch)
+		defer iter.Close()
+
+		// NewCidV1 makes a copy of the multihash buffer, so we can reuse it to
+		// contain allocs.
+		var buf []byte
+		for iter.Rewind(); iter.Valid(); iter.Next() {
+			if ctx.Err() != nil {
+				return // context has fired.
+			}
+			if atomic.LoadInt64(&b.state) != stateOpen {
+				// open iterators will run even after the database is closed...
+				return // closing, yield.
+			}
+			k := iter.Item().Key()
+			if b.prefixing {
+				k = k[b.prefixLen:]
+			}
+
+			if reqlen := base32.RawStdEncoding.DecodedLen(len(k)); len(buf) < reqlen {
+				buf = make([]byte, reqlen)
+			}
+			if n, err := base32.RawStdEncoding.Decode(buf, k); err == nil {
+				ch <- cid.NewCidV1(cid.Raw, buf[:n])
+			} else {
+				log.Warnf("failed to decode key %s in badger AllKeysChan; err: %s", k, err)
+			}
+		}
+	}()
+
+	return ch, nil
+}
+
+// HashOnRead implements Blockstore.HashOnRead. It is not supported by this
+// blockstore.
+func (b *Blockstore) HashOnRead(_ bool) {
+	log.Warnf("called HashOnRead on badger blockstore; function not supported; ignoring")
+}
+
+// PooledStorageKey returns the storage key under which this CID is stored.
+//
+// The key is: prefix + base32_no_padding(cid.Hash)
+//
+// This method may return pooled byte slice, which MUST be returned to the
+// KeyPool if pooled=true, or a leak will occur.
```

Comment:
If it is not returned leak won't really happen but the efficiency of the pool will fall.
It might be interesting to evaluate defering the return of key vs returning it maually in the happy path.